src/Control/Comonad/Cofree/Cofreer.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE FlexibleInstances, GADTs, MultiParamTypeClasses, TypeFamilies, UndecidableInstances #-}
+module Control.Comonad.Cofree.Cofreer
+( CofreerF(..)
+, Cofreer(..)
+, headF
+, tailF
+, cowrap
+, coiter
+, unfold
+, annotating
+, annotatingBidi
+, extract
+, unwrap
+) where
+
+import Control.Arrow ((&&&))
+import Control.Comonad
+import Control.Comonad.Cofree.Class
+import Data.Bifunctor
+import Data.Functor.Classes
+import Data.Functor.Foldable hiding (unfold)
+
+data CofreerF f a b where
+  Cofree :: a -> (x -> b) -> f x -> CofreerF f a b
+
+headF :: CofreerF f a b -> a
+headF (Cofree a _ _) = a
+
+tailF :: Functor f => CofreerF f a b -> f b
+tailF (Cofree _ t r) = t <$> r
+
+
+newtype Cofreer f a = Cofreer { runCofreer :: CofreerF f a (Cofreer f a) }
+
+infixr 5 `cowrap`
+cowrap :: a -> f (Cofreer f a) -> Cofreer f a
+cowrap a r = Cofreer (Cofree a id r)
+
+coiter :: Functor f => (b -> f b) -> b -> Cofreer f b
+coiter f = unfold (id &&& f)
+
+unfold :: Functor f => (b -> (a, f b)) -> b -> Cofreer f a
+unfold f c = let (x, d) = f c in Cofreer (Cofree x (unfold f) d)
+
+
+annotating :: Functor f => (f a -> a) -> f (Cofreer f a) -> Cofreer f a
+annotating algebra base = Cofreer (Cofree (algebra (extract <$> base)) id base)
+
+annotatingBidi :: Functor f => (CofreerF f b a -> a) -> CofreerF f b (Cofreer f a) -> Cofreer f a
+annotatingBidi algebra base = Cofreer (Cofree (algebra (extract <$> base)) id (tailF base))
```

Comment:
These annotating functions allow the reuse of a single algebra to both compute a value and annotate every level of the tree with the corresponding value.